[Flang] Allow Intrinsic simpification with min/maxloc dim and scalar result

davemgreen opened the pull request:
This makes an adjustment to the existing fir minloc/maxloc generation code to handle functions with a dim=1 that produce a scalar result. This should allow us to get the same benefits as the existing generated minmax reductions.

This is a recommit of #76194 with an extra alteration to the end of genRuntimeMinMaxlocBody to make sure we convert the output array to the correct type (a `box<heap<i32>>`, not `box<heap<array<1xi32>>>`) to prevent writing the wrong type of box into it. This still allocates the data as a `array<1xi32>`, converting it into a i32 assuming that is safe. An alternative would be to allocate the data as a i32 and change more of the accesses to it throughout genRuntimeMinMaxlocBody.

psteinfeld reviewed (commented):
I've run our internal tests with this change, and I'm seeing problems.  It will take me a little while to produce sanitized test cases, but please don't merge these changes as is.  I will create the test case or cases promptly.

davemgreen commented:
OK thanks for checking - Maybe this cannot get away with converting the array to an int. Let me put up a new version where it changes the alloc type and if there are still problems we can try and figure out why.

psteinfeld commented:
> OK thanks for checking - Maybe this cannot get away with converting the array to an int. Let me put up a new version where it changes the alloc type and if there are still problems we can try and figure out why.

I've looked at the test results.  There were two problems.  First, one of our tests started issuing warnings at compile time when lowering.  Here's an example that shows the problem:
```
program bug
  real :: x(2) = [ 1.0, 2.0 ]
  call test(x,-1)
contains
  subroutine test(array, dim)
    real,intent(in) :: array(2)
    integer,intent(in) :: dim
    integer :: loc
    loc = minloc(array,dim)
  end subroutine
end program
```
Second,I'm seeing differences in what floating point exceptions get raised when doing IEEE arithmetic.  Here's a program that shows that difference:
```
program bug
  real :: a(3), nan, positive_infinity

  nan = 0.0
  positive_infinity = 0.0
  nan = 0.0/nan
  positive_infinity = 1.0/positive_infinity

  a(3) = positive_infinity
  a(1:2) = nan
  if (minloc (a, dim = 1).ne.3) STOP 5
end
```

davemgreen commented:
> > OK thanks for checking - Maybe this cannot get away with converting the array to an int. Let me put up a new version where it changes the alloc type and if there are still problems we can try and figure out why.
> 
> I've looked at the test results. There were two problems. First, one of our tests started issuing warnings at compile time when lowering. Here's an example that shows the problem:
> 
> ```
> program bug
>   real :: x(2) = [ 1.0, 2.0 ]
>   call test(x,-1)
> contains
>   subroutine test(array, dim)
>     real,intent(in) :: array(2)
>     integer,intent(in) :: dim
>     integer :: loc
>     loc = minloc(array,dim)
>   end subroutine
> end program
> ```
> 
> Second,I'm seeing differences in what floating point exceptions get raised when doing IEEE arithmetic. Here's a program that shows that difference:
> 
> ```
> program bug
>   real :: a(3), nan, positive_infinity
> 
>   nan = 0.0
>   positive_infinity = 0.0
>   nan = 0.0/nan
>   positive_infinity = 1.0/positive_infinity
> 
>   a(3) = positive_infinity
>   a(1:2) = nan
>   if (minloc (a, dim = 1).ne.3) STOP 5
> end
> ```

Oh, thanks for the details. That certainly isn't what I expected the problems to be!

I had looked into the Nan issue a little last week, but did not think to test it with inf. This will be a pre-existing issue that is now starting to happen with dim too, as in a test like the one below. I'll see about getting that fixed.
```
program bug
  real :: a(3), nan, positive_infinity
  integer :: b(1)

  nan = 0.0
  positive_infinity = 0.0
  nan = 0.0/nan
  positive_infinity = 1.0/positive_infinity

  a(3) = positive_infinity
  a(1:2) = nan
  b = minloc (a)
  if (b(1) .ne.3) STOP 5                     
end
```

For the other issue - I can imagine that a previous runtime error now becomes a compile time warning. I don't see the same warning, but it seems to go from a runtime error to just working (there is only really one valid value for dim as far as I understand) and it appears gfortran does the same.  I can maybe make it only work for constant dim values, if that would be better than not producing a runtime error message?

psteinfeld commented:
> For the other issue - I can imagine that a previous runtime error now becomes a compile time warning. I don't see the same warning, but it seems to go from a runtime error to just working (there is only really one valid value for dim as far as I understand) and it appears gfortran does the same. I can maybe make it only work for constant dim values, if that would be better than not producing a runtime error message?

To be clear, the warning is coming from lowering.  We don't want users to get any warnings from lowering.  Here's what I get:
```
warning: loc("/local/home/psteinfeld/main/david/install/bug.f90":5:3): Unhandled parameter attribute 'fir.bindc_name'
warning: loc("/local/home/psteinfeld/main/david/install/bug.f90":5:3): Unhandled parameter attribute 'fir.bindc_name'
```

davemgreen commented:
Which options are you using to get that? Cheers

psteinfeld commented:
> Which options are you using to get that? Cheers

I just typed `flang-new bug.f90`

davemgreen committed:
[Flang] Allow Intrinsic simpification with min/maxloc dim and scalar result.

This makes an adjustment to the existing fir minloc/maxloc generation
code to handle functions with a dim=1 that produce a scalar result. This
should allow us to get the same benefits as the existing generated
minmax reductions.

This is a recommit of #75820 with an extra alteration to the end of
genRuntimeMinMaxlocBody to make sure we convert the output array to the correct
type (a `box<heap<i32>>`, not `box<heap<array<1xi32>>>`) to prevent writing the
wrong type into it. This still allocates the data as a `array<1xi32>`,
converting that into a i32 assuming that is safe. The alternative would be to
allocate the data as a i32 and change more of the accesses to it throughout
genRuntimeMinMaxlocBody.

davemgreen commented:
I've rebased after #82313. As far as I understand the Nan issue should now be fixed. A couple of other questions:
 - Is the invalid Dim index OK to not diagnose at higher optimization levels?
 - Does the conversion of `fir.convert %[[V1]] : (!fir.heap<!fir.array<1xi32>>) -> !fir.heap<i32>` sound OK to perform?

Thanks

tblah commented:
>     * Is it OK not to diagnose the invalid Dim index at higher optimization levels?
So far as I can tell, the standard does not say what should happen for an invalid DIM index, so I suppose this is fine so long as there's a tangible benefit to skipping this check. @psteinfeld do you have an opinion on this?

Are you skipping this check for performance, or is this just because there isn't a convenient way to output an error message and exit from FIR?

>     * Does the conversion of `fir.convert %[[V1]] : (!fir.heap<!fir.array<1xi32>>) -> !fir.heap<i32>` sound OK to perform?
I believe this should be safe.

davemgreen commented:
> So far as I can tell, the standard does not say what should happen for an invalid DIM index, so I suppose this is fine so long as there's a tangible benefit to skipping this check. @psteinfeld do you have an opinion on this?
> 
> Are you skipping this check for performance, or is this just because there isn't a convenient way to output an error message and exit from FIR?

Thanks - It just doesn't naturally check at runtime. Constants should be checked at compile time, as far as I understand. It will be the variable case from https://godbolt.org/z/f86YE1WeM that might not be diagnosed. (I had noticed that gfortran does the same thing, assuming that the dim will be a correct value).

I can add a check at compile time that the dim is a constant and the correct value if that would be more preferable.

tblah reviewed (approved):
Okay I had a look at the runtime library and I don't think that catches a dynamically out of bounds dim either. So I think this is okay. +1 from me

I don't think a compile time check for constants is right here (except maybe as an assertion). This sort of thing should be caught in semantics so we can generate a properly formatted warning/error. What I would have considered was a runtime check for DIM, but as the runtime library code appears not to be doing that, I think it is okay not to do that here either.

Thanks for your work on this

davemgreen commented:
Thanks.

This has been committed twice already. @psteinfeld would you be unhappy with it being committed now? I'd like to avoid it being reverted again if we can avoid it! Cheers.

psteinfeld commented:
> Thanks.
> 
> This has been committed twice already. @psteinfeld would you be unhappy with it being committed now? I'd like to avoid it being reverted again if we can avoid it! Cheers.

I'll retest and post my results.  Stay tuned ....

psteinfeld commented:
> I'll retest and post my results. Stay tuned ....

My testing looks good!  Thanks for sticking with this.

davemgreen commented:
> > I'll retest and post my results. Stay tuned ....
> 
> My testing looks good! Thanks for sticking with this.

Thanks for testing. I appreciate the patience and will attempt not to use so much of your time in the future. Please let me know if anything still goes wrong. Thanks
